README.md
[![Crates.io](https://img.shields.io/crates/v/the_logger?logo=rust)](https://crates.io/crates/the_logger)
![Crates.io](https://img.shields.io/crates/l/the_logger/0.5.0)
[![Rust](https://github.com/tommyHellraiser/the_logger/actions/workflows/compile.yml/badge.svg)](https://github.com/tommyHellraiser/the_logger/actions/workflows/compile.yml)
[![rustc_version]][Rust 1.72]
[![the_logger]][the_logger_docs]


[rustc_version]: https://img.shields.io/badge/rustc_version-1.72+-lightgray.svg
[Rust 1.72]: https://blog.rust-lang.org/2023/08/24/Rust-1.72.0.html
[the_logger]: https://img.shields.io/badge/the__logger-docs.rs-green?logo=docs.rs

[the_logger_docs]: https://docs.rs/the_logger/0.5.3/the_logger/

# the_logger

This crate was created with the purpose of making logging easier to the point where you need only two lines
to create a log file. But besides being easy to use and spawn, it's also really customizable down to the 
microseconds.

This logging system is based on the std library to write to files, and uses tokio only for globally locking
the file writer, lazy_static to keep the static reference of the file writer and chrono to fetch the date
and time for the logs. Optionally (coming in a near future version), you can configure its settings by
creating a json file and setting the config in it.

## Features
- Really easy to call and use.
- Very customizable, with more personalization always coming little by little.
- Lightweight, with only 3 dependencies with the default features (lazy_static, tokio and chrono).
- Accessible anywhere in the code, no need to pass by reference.
- Configurable parameters on startup and on the run.

## How to use
As mentioned, base usage is really simple, no configuration is required to log in its default settings. An async 
function is required to call the logger:
````rust
use the_logger::{log_warning, TheLogger};

async fn init_logger(thread_id: u8) {
    let logger: &TheLogger = TheLogger::instance();
    log_warning!(logger, "This is a warning emitted by thread {}", thread_id);
}
````

As shown in the example above, only 2 lines of code were necessary to set up the logger and log a message. 
There are a couple of ways to execute the logging but the least verbose one is via a macro call. Another 
way of doing it is by:

````rust
use the_logger::TheLogger;

async fn init_logger(thread_id: u8) {
    TheLogger::instance()
        .warning().await
        .log_in_file(
            (file!(), line!(), column!()
        ), "This is another logging example").await;
}
````
The latter way reduces the amount of lines to 1, but at the expense of making the call a lot more verbose and least 
intuitive, missing the purpose of this crate.

As seen in the first example, the log_X!() macros have built-in support for the format!() macro, no need to allocate 
an extra variable to build the message. The first parameter of the macro needs to be the logger element, and the rest
of them can be used as built-in String formatter, supporting all the format!() macro features:

````rust
use the_logger::{log_warning, TheLogger};

async fn init_logger(thread_id: u8) {
    let logger: &TheLogger = TheLogger::instance();
    //  This log...
    log_warning!(logger, "This is a log by thread {}", thread_id);
    
    //  ...is equivalent content-wise to this log
    let msg = format!("This is a log by thread {}", thread_id);
    log_warning!(logger, msg);
}
````

There are 2 main ways to use this logger: passing the instance by reference, or calling an instance from anywhere you 
want to:

````rust
use the_logger::{log_info, TheLogger};

async fn parent_function() {
    let logger: &TheLogger = TheLogger::instance();
    
    child_function_one(logger).await;
    
    child_function_two();
}

async fn child_function_one(logger: &TheLogger) {
    
    log_info!(
        logger, 
        "This will trigger an Informational log by passing the logger instance as reference"
    );
    
}

async fn child_function_two() {
    let logger: &TheLogger = TheLogger::instance();
    
    log_info!(
        logger, 
        "This will trigger an Informational log by calling an instance of the logger without parameters"
    );
}
````
This is possible due to the use of tokio's RwLock feature mentioned at the beginning. This method decided to avoid 
inconsistencies and errors when writing to the log files. This way, no matter how many threads are running, 
the_logger crate will ensure the log lines are written one at a time, and in the order they were called.

## Log files
All the files generated will be placed inside the project using this crate, in the ``logs/`` folder with the 
name: ``"Log <year>-<month>-<day>.log"``

In future versions you'll be able to configure the name (date will always be present for better files organization), the
save path, the file format, and you'll also be able to delete files older than a configurable amount of days.


## Log levels
There are 7 log levels, that aren't all actually levels, but rather categories. These are:
- Verbose
- Informational
- Error
- Warning
- Debug
- Trace
- Critical

Each one of them may serve different purposes in your code, and are present for you to use them as you see fit.

To invoke each one of them, you can simply use the defined macro calls:
````rust
use the_logger::TheLogger;
use the_logger::{log, log_info, log_error, log_warning, log_debug, log_trace, log_critical};

async fn init_logger(thread_id: u8) {
    let logger: &TheLogger = TheLogger::instance();
    
    log!(logger, "This call will trigger a Verbose log");
    log_info!(logger, "This call will trigger an Informational log");
    log_error!(logger, "This call will trigger an Error log");
    log_warning!(logger, "This call will trigger a Warning log");
    log_debug!(logger, "This call will trigger a Debug log");
    log_trace!(logger, "This call will trigger a Trace log");
    log_critical!(logger, "This call will trigger a Critical log");
}
````

## Configuration
As mentioned in the beginning, this logger is very customizable. You can show and hide almost all of its elements:
- Years in the date
- Months in the date
- Days in the date
- Hours in the time
- Minutes in the time
- Seconds in the time
- Milliseconds in the time
- Microseconds in the time
- Log level
- Log location (file, line and column)

The default configuration is to show every of the aforementioned elements, except for the column in the file location
options.

You can also configure the maximum length of the location and content text. Location text content is limited to 60 
characters by default, and the log content itself, to 300.

There are 2 ways of configuring this logger: on startup and on the run (will be adding configuration by json file in 
the future). To configure on startup:
````rust
use the_logger::{TheLogger, TheLoggerConfig};

async fn config_logger() {
    let logger_config = TheLoggerConfig::default()
        .hide_millisecs()               //  This will hide both the milliseconds and the microseconds
        .hide_level()                   //  This will hide the log level
        .hide_location()                //  This will hide the file name, line and column
        .location_content_length(100)   //  This will set the maximum location content text length to 100
        .log_content_length(1000);      //  This will set the maximum log content text length to 1000
    //  Feel free to explore the documentation for more configuration options
    
    let logger: &TheLogger = TheLogger::instance().config_logger(logger_config).await;
    //  The logger is now configured with the options set above
}
````

This configuration method is recommended only at startup, since sending the TheLoggerConfig struct instance to the 
``.config_logger()`` will overwrite the current contents of the configuration. For on-the-go configuration, there's 
a more effective way of doing it:

````rust
use the_logger::{TheLogger, TheLoggerConfig};

async fn config_logger() {
    //  Logger instanced at start
    let logger: &TheLogger = TheLogger::instance();
    
    //  Logger configured on the go
    logger.hide_millisecs().await
        .hide_level().await
        .hide_location().await
        .location_content_length(100).await
        .log_content_length(1000).await;
    
    //  The configuration applied is the same as in the example above
}
````

## Log content examples
In the following examples, every parameter was left to its default vale, except the location content length, which was 
set to 20 characters:
````text
2023-12-17 01:51:10.677106	[VERBOSE]	@src\main.rs: 37    Log from thread 10
2023-12-17 01:51:10.677219	[WARNING]	@src\main.rs: 37    Log from thread 17
2023-12-17 01:51:10.677343	[TRACE]		@src\main.rs: 34    Log from thread 2
2023-12-17 01:51:10.702689	[CRITICAL]	@src\main.rs: 28    Log from thread 0
2023-12-17 01:51:10.810655	[ERROR]		@src\main.rs: 31    Log from thread 1
2023-12-17 01:51:11.367695	[INFO]		@src\main.rs: 34    Log from thread 2
2023-12-17 01:51:11.538744	[DEBUG]		@src\main.rs: 40    Log from thread 4
````

The previous log lines were selected out of approximately 350 lines generated by 30 threads. They were not extracted as 
a block, since some of the threads had longer sleep times than others, causing repetitions, so I had to select some of 
these lines.

The following is an example of a log line with all default settings:
````text
2023-12-17 01:55:43.144632	[WARNING]	@src\main.rs: 43                                            Log from thread 26
````

You can see how it becomes useful to be able to hide certain elements and configure the length of the location content 
and/or the log text content. For the purpose of better readability, we'll use a location content length of 20 chars
moving forward. Now, if we remove the log level we get:
````text
2023-12-17 01:58:57.874425		@src\main.rs: 39    Log from thread 24
````

Next, we'll also remove the milliseconds, which will inherently remove the microseconds as well:
````text
2023-12-17 01:59:57		@src\main.rs: 37    Log from thread 2
````

But you might have a moment by moment tracking of your program or service, and the log date might not be useful, only
the full timestamp, level, location and content, in that case:
````text
02:01:52.509640	[CRITICAL]	@src\main.rs: 32    Log from thread 0
````

You might also have a small program that has a single file and the error is very descriptive, you can configure 
it to get:
````text
02:02:54.830615	[INFO]		Log from thread 16
````

What if you only want the log level and the content because your error already includes date, time, and location? 
You can do that too:
````text
[CRITICAL]	Log from thread 14
````

And finally, if ouy want to hide everything, except the log text content:
````text
Log from thread 25
````

The only parameter you can't hide is the log text content, since it's the minimum requirement for a log.

## Some testing and conclusions

The ability to disable elements to log into the file comes with two major benefits: faster program execution due to the 
reduced amount of data that needs to be written in the file, and related to the amount of data is the file size.

To test the file size, I ran a program with 100 threads for two minutes in the full configuration (default parameters)
and with minimal configuration (only the log text content), and the file sizes were:
- Around 1116KB for the full configuration
- Around 177KB for the minimal configuration

Which means the file size was reduced by approximately 84% by leaving only the minimum required data. Of course, that 
amount of threads introduced bottlenecks in file writing because of the time they had to wait for the lock to be released,
but the size ratio remained constant across the different tests nevertheless. To ensure this was the case, I reduced 
the amount of threads to 30, with the same amount of time, and the results were:
- Around 792KB for the full configuration
- Around 124KB for the minimal configuration

Which gives the same full_size/minimal_size ratio.

## Future features and bug reports
As I mentioned across this readme file, I have in mind some improvements and features, that include:
- Configuration by creating a json file in a folder, yet to be determined
- Configuration of the log file path
- Configuration of the log file name
- Configuration of some separators
- Configuration of the log level enclosure -> [ERROR] vs {ERROR} vs (ERROR) vs ERROR vs |ERROR|, etc

If you have any suggestions, bugs to report, or any other kind of feedback, I'll leave some contact channels ion the 
next section.

## Wrap-up
That sums this crate up, I hope you find it useful, and if you have any questions, suggestions or reports, feel free
to contact me at:

email: ``nacho.ponce25@gmail.com``

telegram: ``@tommyHellraiser``

Or submit an issue in the GitHub repository:
https://github.com/tommyHellraiser/the_logger

Thank you for reading, and have a great one!

changelog.md
# v0.5.0 - Initial release
- TheLogger supports configuration on startup and once it's already been instantiated.
- This logger can be configured to turn off every log element almost individually down to the microseconds if needed, 
except for the log text content (message itself to log).
- It uses tokio to lock the file writer, lazy_static to keep a global instance and chrono to fetch the datetime.
- Only one line to instantiate, and one line to execute a log.
- Support for 7 different log levels:
  - Verbose
  - Informational
  - Error
  - Warning
  - Debug
  - Trance
  - Critical
- Support to configure the length of the log location (file name, line, column) content maximum length, and also the 
log text content maximum length.
- All file writing is done using the std library.
- Global RwLock of the file writer to ensure each thread logs its message in the requested order, and to avoid
inconsistencies and errors when logging from multiple threads.
- Use TheLogger by passing it as a parameter to different functions, or by calling an instance of it.
- It's a singleton, so a single static reference will run through the whole program
- Tested locally with my personal projects, so ANY usage and bug reports will be greatly appreciated.

src/logger/the_logger.rs
use std::fs;
use lazy_static::lazy_static;
use std::fs::File;
use std::io::Write;
use tokio::sync::RwLock;
use crate::logger::logger_config::{LogLevel, TheLoggerConfig};

lazy_static!(
    /// Static reference that allows the user to access the logger from anywhere in the code.
    static ref THE_LOGGER: TheLogger = TheLogger::new();
);

/// Main struct to instantiate when using TheLogger. Its inner RwLock allows only one usage of the file writer and
/// configuration member at a time.
pub struct TheLogger {
    inner: RwLock<TheLoggerInner>
}

#[doc(hidden)]
struct TheLoggerInner {
    config: TheLoggerConfig,
    file_writer: File
}



impl TheLogger {
    #[doc(hidden)]
    fn new() -> Self {
        fs::create_dir_all("./logs/").unwrap();
        Self {
            inner: RwLock::new(TheLoggerInner {
                config: TheLoggerConfig::default(),
                file_writer: std::fs::OpenOptions::new()
                                     .write(true)
                                     .create(true)
                                     .append(true)
                                     .open(
                                         format!("./logs/Log {}.log", chrono::Local::now().naive_local().format("%Y-%m-%d"))
                                     ).unwrap()
            })
        }
    }

    /// ## Description
    /// Returns the instance of the logger. This is the main way to access the logger, and it'll be configured by default.
    ///
    /// TheLogger has built-in support for the format! macro, so the user can use it to format the log message without
    /// having to allocate an extra variable, as shown in the example below.
    ///
    /// Its setup can be changed later using the methods provided by this crate.
    ///
    /// ### Default settings:
    /// - All date elements enabled
    /// - All time elements enabled
    /// - Log level enabled and set to VERBOSE
    /// - Time format set to Local time
    /// - File name and line number shown, but column number is hidden.
    /// - File name, location, column information section is limited to a maximum default of 60 chars, configurable.
    /// - Log content is limited to a maximum default of 300 chars, also configurable.
    ///
    /// ### Log Example
    /// ```text
    /// 2023-12-16 17:08:07.451851800  [VERBOSE]  This is a log example for the the_logger crate
    /// ```
    ///
    /// ### Usage example
    /// ````rust
    /// use the_logger::{log_warning, TheLogger};
    ///
    /// async fn init_logger(thread_id: u8) {
    ///     let logger: &TheLogger = TheLogger::instance();
    ///     log_warning!(logger, "This is a warning emitted by thread {}", thread_id);
    /// }
    /// ````
    pub fn instance() -> &'static Self {
        &THE_LOGGER
    }

    /// ## Description
    /// Executes the logging to the file according to the current configuration
    pub async fn log_in_file(&self, (file, line, column): (&str, u32, u32), incoming_msg: &str) {
        let mut msg = String::new();
        let mut location_info = String::new();
        let location_length;
        let content_length;

        {
            let inner = self.inner.read().await;

            //  Datetime formatting
            let datetime = if inner.config.get_utc_config() {
                chrono::Utc::now().naive_utc()
            } else {
                chrono::Local::now().naive_local()
            };
            let mut datetime_format = String::new();

            //  Date formatting
            let mut space_date_time = false;
            let mut time_shown = false;
            if !inner.config.get_years_config() {
                datetime_format.push_str("%Y");
                space_date_time = true;
            }
            if !inner.config.get_months_config() {
                space_date_time = true;
                datetime_format.push_str("-%m");
            }
            if !inner.config.get_days_config() {
                space_date_time = true;
                datetime_format.push_str("-%d");
            }
            if space_date_time {
                datetime_format.push(' ');
            }

            //  Time formatting
            if !inner.config.get_hours_config() {
                datetime_format.push_str("%H");
                time_shown = true;
            }
            if !inner.config.get_minutes_config() {
                datetime_format.push_str(":%M");
                time_shown = true;
            }
            if !inner.config.get_seconds_config() {
                datetime_format.push_str(":%S");
                time_shown = true;
            }
            match (inner.config.get_millisecs_config(), inner.config.get_microsecs_config()) {
                (false, false) => {
                    datetime_format.push_str(".%6f");
                    time_shown = true;
                },
                (false, true) => {
                    datetime_format.push_str(".%3f");
                    time_shown = true;
                },
                (true, false) => {},
                _ => {}
            }
            if space_date_time || time_shown {
                datetime_format.push('\t');
            }
            msg.push_str(&datetime.format(&datetime_format).to_string());

            //  Log level type config
            if !inner.config.get_level_config() {
                match inner.config.get_log_level() {
                    LogLevel::Verbose => {
                        msg.push_str("[VERBOSE]\t");
                    },
                    LogLevel::Information => {
                        msg.push_str("[INFO]\t\t");
                    },
                    LogLevel::Error => {
                        msg.push_str("[ERROR]\t\t");
                    },
                    LogLevel::Warning => {
                        msg.push_str("[WARNING]\t");
                    },
                    LogLevel::Debug => {
                        msg.push_str("[DEBUG]\t\t");
                    },
                    LogLevel::Trace => {
                        msg.push_str("[TRACE]\t\t");
                    },
                    LogLevel::Critical => {
                        msg.push_str("[CRITICAL]\t");
                    }
                }
                //  Only insert tab if location data is not shown
            } else if !inner.config.get_file_name_config() || !inner.config.get_file_line_config() || inner.config.get_file_column_config() {
                msg.push('\t');
            }

            //  File location configuration
            if !inner.config.get_file_name_config() {
                location_info.push_str(format!("@{}", file).as_str());

                if !inner.config.get_file_line_config() {
                    location_info.push_str(format!(": {}", line).as_str());

                    if inner.config.get_file_column_config() {
                        location_info.push_str(format!("|{}", column).as_str());
                    }
                }
            }

            //  Location info and content lengths configuration
            location_length = inner.config.get_location_length();
            content_length = inner.config.get_log_content_length();
        }

        //  Close the message string and log it. Trim the message if it's longer than the configured lengths
        if location_length < location_info.len() {
            location_info = location_info.as_str()[0..location_length].to_string();
            //  Append a tab to separate the location info and the content a little bit
            location_info.push_str("\t\t");
        }
        if !location_info.is_empty() {
            msg.push_str(format!("{:<location_length$}", location_info).as_str());
        }
        if content_length < incoming_msg.len() {
            msg.push_str(incoming_msg[0..content_length].as_ref());
        } else {
            msg.push_str(incoming_msg);
        }
        self.inner.write().await.file_writer.write_all(format!("{}\n", msg).as_bytes()).unwrap();
    }

    /// ## Description
    /// Allows the user to configure the logger using a single config call. The configuration must previously be created
    /// by instantiating the TheLoggerConfig struct adn setting each desired field with its builder methods.
    ///
    /// Every change to the configuration is made through this method.
    ///
    /// ### Initial configuration example
    /// ```rust
    /// use the_logger::{TheLogger, TheLoggerConfig};
    ///
    /// async fn init_logger() {
    ///     let logger_config = TheLoggerConfig::default()
    ///         .hide_file_line()
    ///         .hide_microsecs()
    ///         .hide_level();
    ///
    ///     let logger: &TheLogger = TheLogger::instance().config(logger_config).await;
    /// }
    /// ```
    ///
    /// As a result, the user has configured the logger to hide the file line number, the microsecs and the log level,
    ///
    /// and leave the rest of the parameters with their default values.
    ///
    /// ### Modifying configuration example
    /// If the user needs to modify the logger configuration for any particular reason, for example for a specific
    ///
    /// routine that doesn't need to show all the data or needs to show different data, this can be achieved in a similar
    ///
    /// way to the initial configuration:
    ///
    /// ```rust
    /// use the_logger::{TheLogger, TheLoggerConfig};
    ///
    /// async fn config_logger() {
    ///     let logger_config = TheLoggerConfig::default()
    ///         .show_file_line()
    ///         .show_microsecs()
    ///         .show_level();
    ///
    ///     let logger: &TheLogger = TheLogger::instance().config(logger_config).await;
    /// }
    /// ```
    ///
    /// As opposed to the previous example, we're now reverting the changes made to the logger configuration, so that
    ///
    /// the file line number, microseconds stamp and log level are shown.
    pub async fn config(&self, logger_config: TheLoggerConfig) -> &Self {
        self.inner.write().await.config = logger_config;
        self
    }

    //////////////////////
    /* Log Type methods */
    //////////////////////
    /// ## Description
    /// Configures the log level as verbose adding the [[VERBOSE]] tag
    pub async fn verbose(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Verbose);
        self
    }

    /// ## Description
    /// Configures the log level as informational adding the [[INFO]] tag
    pub async fn info(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Information);
        self
    }

    /// ## Description
    /// Configures the log level as error adding the [[ERROR]] tag
    pub async fn error(&self) -> &Self{
        self.inner.write().await.config.set_log_level(LogLevel::Error);
        self
    }

    /// ## Description
    /// Configures the log level as warning adding the [[WARNING]] tag
    pub async fn warning(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Warning);
        self
    }

    /// ## Description
    /// Configures the log level as debug adding the [[DEBUG]] tag
    pub async fn debug(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Debug);
        self
    }

    /// ## Description
    /// Configures the log level as trace adding the [[TRACE]] tag
    pub async fn trace(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Trace);
        self
    }

    /// ## Description
    /// Configures the log level as critical adding the [[CRITICAL]] tag
    pub async fn critical(&self) -> &Self {
        self.inner.write().await.config.set_log_level(LogLevel::Critical);
        self
    }

    ///////////////////////////
    /* Configuration methods */
    ///////////////////////////
    /* Hide methods */
    //////////////////
    /// ## Description
    /// Configures the log date to hide the years. Default is to show them
    pub async fn hide_years(&self) -> &Self {
        self.inner.write().await.config.set_years_config(true);
        self
    }

    /// ## Description
    /// Configures the log date to hide the months. Default is to show them
    pub async fn hide_months(&self) -> &Self {
        self.inner.write().await.config.set_months_config(true);
        self
    }

    /// ## Description
    /// Configures the log date to hide the days. Default is to show them
    pub async fn hide_days(&self) -> &Self {
        self.inner.write().await.config.set_days_config(true);
        self
    }

    /// ## Description
    /// Configures the log time to hide the hours. Default is to show them
    pub async fn hide_hours(&self) -> &Self {
        self.inner.write().await.config.set_hours_config(true);
        self
    }

    /// ## Description
    /// Configures the log time to hide the minutes. Default is to show them
    pub async fn hide_minutes(&self) -> &Self {
        self.inner.write().await.config.set_minutes_config(true);
        self
    }

    /// ## Description
    /// Configures the log time to hide the seconds. Default is to show them
    pub async fn hide_seconds(&self) -> &Self {
        self.inner.write().await.config.set_seconds_config(true);
        self
    }

    /// ## Description
    /// Configures the log time to hide the milliseconds. Default is to show them
    ///
    /// ### Warning
    /// Hiding the milliseconds and showing the microseconds would cause an unexpected time tracking in the logs,
    ///
    /// therefore, in this specific case, both milliseconds and microseconds will be hidden
    pub async fn hide_millisecs(&self) -> &Self {
        self.inner.write().await.config.set_millisecs_config(true);
        self
    }

    /// ## Description
    /// Configures the log time to hide the microseconds. Default is to show them
    ///
    /// ### Warning
    /// Hiding the milliseconds and showing the microseconds would cause an unexpected time tracking in the logs,
    ///
    /// therefore, in this specific case, both milliseconds and microseconds will be hidden
    pub async fn hide_microsecs(&self) -> &Self {
        self.inner.write().await.config.set_microsecs_config(true);
        self
    }

    /// ## Description
    /// Configures the log timezone to UTC format. Default is Local time
    pub async fn utc_time(&self) -> &Self {
        self.inner.write().await.config.set_utc_config(true);
        self
    }

    /// ## Description
    /// Configures the log level to be hidden. Default is to show it
    pub async fn hide_level(&self) -> &Self {
        self.inner.write().await.config.set_level_config(true);
        self
    }

    /// ## Description
    /// Configures the log file name, line and column to be hidden. Default is to show them
    pub async fn hide_file_name(&self) -> &Self {
        self.inner.write().await.config.set_file_name_config(true);
        self.inner.write().await.config.set_file_line_config(true);
        self.inner.write().await.config.set_file_column_config(false);
        self
    }

    /// ## Description
    /// Configures the log file line and column to be hidden. Default is to show them
    pub async fn hide_file_line(&self) -> &Self {
        self.inner.write().await.config.set_file_line_config(true);
        self
    }

    /// ## Description
    /// Configures the log file column to be shown. Default is to hide it
    pub async fn show_file_column(&self) -> &Self {
        self.inner.write().await.config.set_file_column_config(true);
        self
    }

    //////////////////
    /* Show methods */
    //////////////////
    /// ## Description
    /// Configures the log date to show the years. Default is to show them
    pub async fn show_years(&self) -> &Self {
        self.inner.write().await.config.set_years_config(false);
        self
    }

    /// ## Description
    /// Configures the log date to show the months. Default is to show them
    pub async fn show_months(&self) -> &Self {
        self.inner.write().await.config.set_months_config(false);
        self
    }

    /// ## Description
    /// Configures the log date to show the days. Default is to show them
    pub async fn show_days(&self) -> &Self {
        self.inner.write().await.config.set_days_config(false);
        self
    }

    /// ## Description
    /// Configures the log time to show the hours. Default is to show them
    pub async fn show_hours(&self) -> &Self {
        self.inner.write().await.config.set_hours_config(false);
        self
    }

    /// ## Description
    /// Configures the log time to show the minutes. Default is to show them
    pub async fn show_minutes(&self) -> &Self {
        self.inner.write().await.config.set_minutes_config(false);
        self
    }

    /// ## Description
    /// Configures the log time to show the seconds. Default is to show them
    pub async fn show_seconds(&self) -> &Self {
        self.inner.write().await.config.set_seconds_config(false);
        self
    }

    /// ## Description
    /// Configures the log time to show the milliseconds. Default is to show them
    ///
    /// ### Warning
    /// Hiding the milliseconds and showing the microseconds would cause an unexpected time tracking in the logs,
    ///
    /// therefore, in this specific case, both milliseconds and microseconds will be hidden
    pub async fn show_millisecs(&self) -> &Self {
        self.inner.write().await.config.set_millisecs_config(false);
        self
    }

    /// ## Description
    /// Configures the log time to show the microseconds. Default is to show them
    ///
    /// ### Warning
    /// Hiding the milliseconds and showing the microseconds would cause an unexpected time tracking in the logs,
    ///
    /// therefore, in this specific case, both milliseconds and microseconds will be hidden
    pub async fn show_microsecs(&self) -> &Self {
        self.inner.write().await.config.set_microsecs_config(false);
        self
    }

    /// ## Description
    /// Configures the log timezone to Local format. Default is Local time
    pub async fn local_time(&self) -> &Self {
        self.inner.write().await.config.set_utc_config(false);
        self
    }

    /// ## Description
    /// Configures the log level to be shown. Default is to show it
    pub async fn show_level(&self) -> &Self {
        self.inner.write().await.config.set_level_config(false);
        self
    }

    /// ## Description
    /// Configures the log file name, line and column to be shown. Default is to show them
    pub async fn show_file_name(&self) -> &Self {
        self.inner.write().await.config.set_file_name_config(false);
        self
    }

    /// ## Description
    /// Configures the log file line and column to be shown. Default is to show them
    pub async fn show_file_line(&self) -> &Self {
        self.inner.write().await.config.set_file_line_config(false);
        self
    }

    /// ## Description
    /// Configures the log file column to be shown. Default is to hide it
    pub async fn hide_file_column(&self) -> &Self {
        self.inner.write().await.config.set_file_column_config(false);
        self
    }

    ///////////////////////////
    /* Length configurations */
    ///////////////////////////
    /// ## Description
    /// Configures the log file name, line and column location content's length. Default is 100 characters
    pub async fn location_content_length(&self, length: usize) -> &Self {
        self.inner.write().await.config.set_location_length(length);
        self
    }

    /// ## Description
    /// Configures the log message content's length. Default is 300 characters
    pub async fn log_content_length(&self, length: usize) -> &Self {
        self.inner.write().await.config.set_log_content_length(length);
        self
    }
}
